va/va_prot.h

```diff
@@ -0,0 +1,361 @@
+ * create a VA buffer first with type VAProtectedSessionTeeExecBufferType, the structure of
+ * this type is IHV's implementation specific. Usually it will have a function_id to specify
+ * which function to execute. The following pseudo-code demonstrates getting session id via
+ * vaProtectedSessionTeeExec() as an example.
+ *
+ * \code
+ * uint16_t version = 0xFF;
+ * VABufferID buffer;
+ * VAProtectedSessionTeeExecBuffer tee_exec_buff = {0};
+ *
+ * tee_exec_buff.function_id = GET_TEE_VERSION
+ * tee_exec_buff.input.input_data = nullptr;
+ * tee_exec_buff.input.data_size = 0;
+ * tee_exec_buff.output.output_data = &version;
+ * tee_exec_buff.output.max_data_size = (uint16_t);
+ * va_status = vaCreateBuffer(
+ *                 va_dpy,
+ *                 protected_session,
+ *                 (VABufferType) VAProtectedSessionTeeExecBufferType,
+ *                 sizeof(tee_exec_buff),
+ *                 1,
+ *                 &tee_exec_buff,
+ *                 &buffer);
+ *
+ * va_status = vaProtectedSessionTeeExec(va_dpy, protected_session, buffer);
+ *
+ * vaDestroyBuffer(va_dpy, buffer);
+ * \endcode
+ */
+
+/**
+ * ProtectedSessions and Contexts
+ *
+ * Context represents a "virtual" video decode, encode or video processing
+ * pipeline. Surfaces are render targets for a given context. The data in the
+ * surfaces are not accessible to the client except if derived image is supported
+ * and the internal data format of the surface is implementation specific.
+ *
+ * A protected session is a session between application and GPU hardware.
+ * Protected sessions can be attached to display to enable protection in the pipeline.
+ *
+ * Both contexts and protected sessions are identified by unique IDs and its
+ * implementation specific internals are kept opaque to the clients
+ */
+typedef VAContextID VAProtectedSessionID;
+
+/** \brief Input buffer of VAProtectedSessionTeeExecBuffer or VAProtectedSessionHwUpdateBuffer */
+typedef struct _VAProtectedFuncionBufferInput
+{
+    /** \brief Contains the size of the input data. */
+    uint32_t    data_size;
+    /** \brief The input data. */
+    void        *input_data;
+} VAProtectedFuncionBufferInput;
+
+/** \brief Output buffer of VAProtectedSessionTeeExecBuffer or VAProtectedSessionHwUpdateBuffer */
+typedef struct _VAProtectedFuncionBufferOutput
+{
+    /** \brief The maximum size of data that the driver can return in the output buffer. */
+    uint32_t    max_data_size;
+    /** \brief Returns the size of the output data written by the driver. */
+    uint32_t    data_size;
+    /** \brief Output data written by the driver. */
+    void        *output_data;
+} VAProtectedFuncionBufferOutput;
+
+/** \brief Buffer for vaProtectedSessionTeeExec() */
+typedef struct _VAProtectedSessionTeeExecBuffer {
+    /** \brief Specify the function to execute. It is IHV's implementation specific */
+    uint32_t                        function_id;
```

Comment:
What is a function ID? Is that vendor specific or you plan to standardize them?

Reply:
Yes, this could be vendor specific. For HDCP reference implementation, I define a few function ids for it in
[here](https://github.com/intel/libva/pull/457/commits/68bebef3de18317dcf827ddac73251be2caef068#diff-b511c09906306779cb58f5e7065beb25fc2cd261d72307b0b872ad8382581f9bR264)

Reply:
From the declaration you pointed to I don't feel they are vendor, i.e. Intel specific:
```
    VA_PROTECTED_LINK_EXECUTE_FID_GET_STATUS        = 0x20000000,
```

You basically declared VAAPI level cross-vendor function IDs... (And I don't see how they relate to HDCP as well by the way...)

Reply:
OK, I removed these definitions.